src/idpi/operators/lateral_operators.py

```diff
@@ -0,0 +1,192 @@
+"""Lateral operators."""
+
+# Standard library
+from typing import Literal
+
+# Third-party
+import numpy as np
+import xarray as xr
+
+
+def compute_weights(
+    window_size: int,
+    window_type: Literal["exp", "const"],
+    window_shape: Literal["disk", "square"],
+) -> xr.DataArray:
+    """Compute weights for a convolution kernel.
+
+    Parameters
+    ----------
+    window_size : int
+        Number of grid cells the kernel. Must be an odd number.
+    window_type : {"exp", "const"}
+        The window type.
+
+        Supported types:
+         - exp: Exponential drop from the center of the window.
+         - const: All weights are set to 1.
+
+    window_shape : {"disk", "square"}
+        The window shape.
+
+        Supported shapes:
+         - disk: The kernel is confined to the circle inscribing the window.
+         - square: The kernel is defined on the entire window.
+
+    Raises
+    ------
+    ValueError
+        if `window_size` is not an odd number.
+
+    Returns
+    -------
+    xarray.DataArray
+        Weights corresponding to the `window_type` within the given `window_shape`
+        and zero outside of the shape.
+
+    """
+    n = window_size
+
+    if n % 2 != 1:
+        raise ValueError("window_size must be an odd number.")
+
+    radius = n // 2
+    yy, xx = np.mgrid[:n, :n] - radius
+    dist = np.sqrt(xx**2 + yy**2)
+
+    if window_type == "exp":
+        kernel = np.exp(-dist)
+    else:
+        kernel = np.ones((n, n))
+
+    weights = xr.DataArray(kernel, dims=["win_x", "win_y"])
+
+    if window_shape == "disk":
+        return weights.where(dist <= radius, 0.0)
+
+    return weights
+
+
+def compute_cond_mask(
+    windows: xr.DataArray, weights: xr.DataArray, frac_val: float, nx: int, ny: int
+) -> xr.DataArray:
+    """Compute the conditional mask for which the convolution results are valid.
+
+    The function computes the ratio of undefined values to the total number values
+    in the convolution window. The computational domain and the window shape are both
+    taken into account. The mask is true for locations where the ratio is above the
+    given threshold.
+
+    Parameters
+    ----------
+    windows : xarray.DataArray
+        Values of the constructed windows for the application of the convolution.
+    weights : xarray.DataArray
+        Weights of the convolution kernel.
+    frac_val : float
+        Threshold on the ratio of undefined values.
+    nx : int
+        Size of the field in the x dimension.
+    ny : int
+        Size of the field in the y dimension.
+
+    Returns
+    -------
+    xarray.DataArray
+        Boolean mask indicating where the undefined value ratio in the convolution
+        kernel is within the acceptable threshold.
+
+    """
+    mask = weights > 0
+
+    loc_x = windows.x + windows.win_x - windows.sizes["win_x"] // 2
```

Comment:
this block is hard to follow, not sure if some comments would help to digest it better?

Reply:
For each window, `x` is the coordinate of the center of the window in the parent field. `win_x` is the local coordinate of the current element in the window. `loc_x` represents the location of the window element in the coordinates of the parent field. This is needed to filter out nans that are introducing by the padding of the field by the rolling method. Only the nans that are present within the bound of the field should count towards the `frac_val` threshold.

Reply:
Should I add this paragraph as a comment?